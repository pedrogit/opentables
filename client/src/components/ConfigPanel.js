import React from "react";
import Box from '@mui/material/Box';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import Stack from "@mui/material/Stack";
import Collapse from "@mui/material/Collapse";
import { useTheme } from "@mui/material/styles";
import IconButton from '@mui/material/IconButton';
import HighlightOffIcon from '@mui/icons-material/HighlightOff';
import PropTypes from 'prop-types';

// local imports
import List from "./List";
const Globals = require("../../../common/globals");
const Schema = require("../../../common/schema");

function TabPanel(props) {
  const { children, value, index, ...other } = props;

  return (
    <div
      role="tabpanel"
      hidden={value !== index}
      id={`simple-tabpanel-${index}`}
      aria-labelledby={`simple-tab-${index}`}
      {...other}
    >
      {value === index && (
        <>
          {children}
        </>
      )}
    </div>
  );
}

TabPanel.propTypes = {
  children: PropTypes.node,
  index: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired,
};

function ConfigPanel({
  showConfigPanel,
  setShowConfigPanel,
  view,
  setViewData,
  setAuthAPIRequest,
  setErrorMsg,
  handleReload
}) {
  const [value, setValue] = React.useState(0);
  const theme = useTheme();

  const handleChange = (event, newValue) => {
    setValue(newValue);
  };

  const handleEditView = React.useCallback(
    (editedView) => {
      var child = {...view[Globals.childlistFieldName]};
      setViewData({
        ...editedView[Globals.childlistFieldName][Globals.itemsFieldName][0],
        [Globals.childlistFieldName]: child
      });
    }, [view, setViewData]
  );

  const handleEditList = React.useCallback(
    (editedList) => {
      var newViewData = {
        ...view,
        [Globals.childlistFieldName]: {
          ...editedList[Globals.childlistFieldName][Globals.itemsFieldName][0],
        }
      }
      if (view[Globals.childlistFieldName][Globals.itemsFieldName]) {
        newViewData[Globals.childlistFieldName][Globals.itemsFieldName] = [...view[Globals.childlistFieldName][Globals.itemsFieldName]];
      }
      setViewData(newViewData);
    }, [view, setViewData]
  );

  var parentViewView = {
    [Globals.childlistFieldName]: {
      ...Globals.listOfAllViews,
      [Globals.itemsFieldName]: view ? [view] : []
    }
  }

  var parentListView = {
    [Globals.childlistFieldName]: {
      ...Globals.listOfAllLists,
      [Globals.itemsFieldName]: view[Globals.childlistFieldName] ? [view[Globals.childlistFieldName]] : []
    }
  }

  var noListRPerm = parentListView[Globals.childlistFieldName][Globals.itemsFieldName][0] === Globals.permissionDeniedOnListOrItems;

  return (
    <Stack sx={{backgroundColor: theme.palette.primary.palebg}}>
      <Collapse in={showConfigPanel}>
        <Stack sx={{
          position: 'relative', 
          borderBottomWidth: '5px', 
          borderBottomStyle: 'solid', 
          borderBottomColor: theme.palette.primary.main
        }}>
          <IconButton
            sx = {{
              position: 'absolute', 
              top: '1px', 
              right: '1px',
              p: theme.openTable.buttonPadding,
              zIndex: '100'
            }}
            id="closeConfigPanelButton"
            aria-label="close error panel" 
            color="inherit"
            onClick={() => setShowConfigPanel(false)}
          >
            <HighlightOffIcon />
          </IconButton>

          <Box sx={{ width: '100%' }}>
            <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
              <Tabs value={value} onChange={handleChange} aria-label="basic tabs example">
                <Tab id="viewProperties" label={Globals.viewProperties}/>
                <Tab 
                  id="listProperties"
                  disabled={noListRPerm}
                  label={Globals.listProperties + (noListRPerm ? " (" + Globals.permissionDenied + ")" : "")}
                />
              </Tabs>
            </Box>
            <TabPanel value={value} index={0}>
              {showConfigPanel && <List
                listType={Globals.viewListType}
                view={parentViewView}
                parsedSchema={new Schema((Globals.listOfAllViews)[Globals.listSchemaFieldName])}
                listSchemaStr={view[Globals.childlistFieldName][Globals.listSchemaFieldName]}
                setAuthAPIRequest={setAuthAPIRequest}
                setViewData={handleEditView}
                setErrorMsg={setErrorMsg}
                showDeleteButton={false}
                handleReload={handleReload}
              />}
            </TabPanel>
            <TabPanel value={value} index={1}>
              {showConfigPanel && <List
                listType={Globals.listListType}
                view={parentListView}
                parsedSchema={new Schema((Globals.listOfAllLists)[Globals.listSchemaFieldName])}
                setAuthAPIRequest={setAuthAPIRequest}
                setViewData={handleEditList}
                setErrorMsg={setErrorMsg}
                showDeleteButton={false}
                handleReload={handleReload}
              />}
            </TabPanel>
          </Box>
        </Stack>
      </Collapse>
    </Stack>
  )
}

export default ConfigPanel;
