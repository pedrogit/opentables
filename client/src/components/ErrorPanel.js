import React from "react";
import Stack from "@mui/material/Stack";
import Alert from "@mui/material/Alert";
import AlertTitle from "@mui/material/AlertTitle";
import { useTheme } from "@mui/material/styles";
import Collapse from "@mui/material/Collapse";
import IconButton from '@mui/material/IconButton';
import HighlightOffIcon from '@mui/icons-material/HighlightOff';

function ErrorPanel({ errorMsg, setErrorMsg, autoClose, closeButton }) {
  const theme = useTheme();
  var closeTimeOutRef = React.useRef(null);
  var openTimeOutRef = React.useRef(null);

  React.useEffect(() => {
    if (errorMsg !== null && errorMsg.text && (!errorMsg.open || errorMsg.open === false)) {
      clearTimeout(openTimeOutRef.current);
      openTimeOutRef.current = setTimeout(() => {
        setErrorMsg({...errorMsg, open: true});
        if (autoClose) {
          clearTimeout(closeTimeOutRef.current);
          closeTimeOutRef.current = setTimeout(() => setErrorMsg({open: false}), 4000);
        }
      }, 300);
    }
  }, [errorMsg, autoClose, setErrorMsg]);

  return (
    <Stack sx={{position: 'relative'}}>
      <Collapse in={errorMsg && errorMsg.open && errorMsg.open === true}>
      <Stack sx={{
        position: 'relative', 
        borderBottomWidth: '5px', 
        borderBottomStyle: 'solid', 
        borderBottomColor: theme.palette.primary.main
      }}>
        {closeButton &&
          <IconButton
            sx = {{position: 'absolute', top: '1px', right: '1px'}}
            id="closeErrorMsgButton"
            aria-label="closeErrorMsgButton" 
            color="inherit"
            onClick={() => setErrorMsg({open: false})}
          >
            <HighlightOffIcon />
          </IconButton>
        }
        <Alert 
            severity={errorMsg && errorMsg.severity ? errorMsg.severity : 'error'} 
            color="primary"
            sx={{padding: '5px'}}
        >
        <AlertTitle>{errorMsg && errorMsg.title ? errorMsg.title : 'Error'}</AlertTitle>
        {(errorMsg && errorMsg.text && errorMsg.open && errorMsg.open === true) ? errorMsg.text : <span> </span>}
        </Alert>
        </Stack>
      </Collapse>
    </Stack>
  )
}

ErrorPanel.defaultProps = {
  autoClose: true,
  closeButton: false
}

export default ErrorPanel;
