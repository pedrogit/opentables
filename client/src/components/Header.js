import React from "react";
import AppBar from "@mui/material/AppBar";
import Toolbar from "@mui/material/Toolbar";
import Box from "@mui/material/Box";
import IconButton from '@mui/material/IconButton';
import AddCircleOutlineIcon from '@mui/icons-material/AddCircleOutline';
import SettingsIcon from '@mui/icons-material/Settings';
import Stack from "@mui/material/Stack";
import Typography from "@mui/material/Typography";

import {LoginButton} from "./LoginForm";

function Header({ 
  viewOwner, 
  viewName, 
  setLoginState, 
  handleOpenConfigPanel, 
  handleAddItem 
}) {
  const [showButtons, setShowButtons] = React.useState(false);
  var buttonsTimeOut = null;

  return (
    <AppBar
      id="headerButtons"
      position="static" 
      onMouseEnter={() => {
        clearTimeout(buttonsTimeOut);
        setShowButtons(true);
      }}
      onMouseLeave={() => {
        buttonsTimeOut = setTimeout(() => setShowButtons(false), 500)
      }}
    >
    <Toolbar variant="dense" disableGutters sx={{ml:'5px'}}>
    {(viewOwner && viewName ? (
        <Stack direction={{ xs: 'column', sm: 'row' }}>
        <Typography sx={{color: 'black', fontStyle: 'italic', mr: '5px'}}>{(viewOwner + '\'s')}</Typography>
        <Typography sx={{fontWeight:'bold'}}>{viewName}</Typography>
        </Stack>
        ) : (
        <Typography>Loading...</Typography>
        ))
    }
    <Box sx={{ flexGrow: 1 }} />
    {showButtons && 
        <Stack direction='row'>
        <LoginButton setLoginState={setLoginState}>Login</LoginButton>
        <IconButton
            id="configPanelOpenButton" 
            aria-label="configPanel" 
            color="inherit"
            onClick={handleOpenConfigPanel}
        >
            <SettingsIcon />
        </IconButton>
        <IconButton
            id="addItemButton" 
            aria-label="addItem" 
            color="inherit"
            onClick={handleAddItem}
        >
            <AddCircleOutlineIcon />
        </IconButton>
        </Stack>
    }
    </Toolbar>
    </AppBar>
  )
}

export default Header;
