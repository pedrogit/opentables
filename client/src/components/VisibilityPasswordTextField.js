import React from "react";
import { TextField, InputAdornment, IconButton } from "@mui/material";
import VisibilityIcon from "@mui/icons-material/Visibility";
import VisibilityOffIcon from "@mui/icons-material/VisibilityOff";
import Tooltip from "@mui/material/Tooltip";

function VisibilityPasswordTextField({ ...rest }) {
  const [isVisible, toggleVisible] = React.useState(false);

  const defaultSx = { backgroundColor: "white", paddingRight: "0px" };

  const {
    name,
    value,
    label,
    required,
    sx,
    inputProps,
    variant,
    size,
    fullWidth,
    autoComplete,
    inputRef,
    onChange,
    onKeyDown,
    error,
    InputLabelProps,
  } = rest;
  return (
    <TextField
      name={name}
      value={value}
      label={label}
      required={required}
      sx={sx}
      inputProps={inputProps}
      variant={variant}
      size={size}
      type={isVisible ? "text" : "password"}
      fullWidth={fullWidth}
      autoComplete={autoComplete}
      onChange={onChange}
      onKeyDown={onKeyDown}
      error={error}
      inputRef={inputRef}
      InputLabelProps={InputLabelProps}
      // eslint-disable-next-line react/jsx-no-duplicate-props
      InputProps={{
        id: "passwordInput",
        sx: { ...defaultSx, ...rest.sx, marginTop: "0px", marginBottom: "0px" },
        endAdornment: (
          <InputAdornment position="end">
            <Tooltip title="View Password">
              <IconButton
                aria-label="Toggle password visibility"
                onClick={() => {
                  toggleVisible(!isVisible);
                }}
                onMouseDown={(event) => {
                  event.preventDefault();
                }}
              >
                {isVisible ? <VisibilityIcon /> : <VisibilityOffIcon />}
              </IconButton>
            </Tooltip>
          </InputAdornment>
        ),
      }}
    />
  );
}

export default VisibilityPasswordTextField;
