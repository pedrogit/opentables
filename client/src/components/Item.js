import React from "react";
//import React, {memo} from "react";
import JsxParser from "react-jsx-parser";
import Stack from "@mui/material/Stack";
import Box from "@mui/material/Box";
import { useTheme } from "@mui/material/styles";
import IconButton from '@mui/material/IconButton';
import HighlightOffIcon from '@mui/icons-material/HighlightOff';

import * as Components from "./components";
const Globals = require("../common/globals");

// Define all components in the current scope
// so they can be used directly, without
// namespace (e.g. Components.Text), in the template
for (let name in Components) {
  global[name] = Components[name];
}

function Item({ 
  template, 
  item, 
  rowNb, 
  setLoginState, 
  handleListAuth, 
  handleDeleteItem,
  setViewId
}) {
  const [newItem, setItem] = React.useState(item);
  const [showButtons, setShowButtons] = React.useState(false);

  const theme = useTheme();

  var defaultSx = {
    bgcolor: rowNb % 2 ? "inherit" : "#EEE", 
    padding: 1,
    position: 'relative'
  };

  var buttonSx = {
    position: 'absolute',
    top: '1px',
    right: '1px'
  }

  var handlePatch = function (val, callback) {
    setLoginState({
      open: false,
      msg: {
        severity: "warning",
        title: "Permission denied",
        text:
        'You do not have permissions to edit "' +
        Object.keys(val)[0] +
        '". Please login with valid credentials...',
      },
      action: {
        method: "patch",
        url: "http://localhost:3001/api/opentables/" + newItem[Globals.itemIdFieldName],
        data: val,
        callback: (success, data) => {
          if (success) {
            setItem(data);
            callback(success, data[Object.keys(val)[0]]);
          }
        }
      },
      tryFirst: true
    });
    return false;
  };

  var handleItemAuth = function(action = 'patch', propName, callback) {
    handleListAuth(action = 'patch', item, propName, callback);
  }

  var setBindings = function (item) {
    //console.log('Item setBindings()...');

    // add property name and rest handlers
    var result = {};
    for (var key in item) {
      if (item.hasOwnProperty(key)) {
        result[key] = {
          handlePatch: handlePatch,
          handleItemAuth: handleItemAuth,
          setViewId: setViewId,
          prop: key,
          val: item[key] ? item[key] : "",
        };
      }
    }
    //result.key = result._id.val;
    return result;
  };



  //console.log('Render Item (' +  item.name + ')...');

  return (
    <Box 
      id={"item_" + (newItem ? newItem[Globals.itemIdFieldName] : '')}
      sx={defaultSx}
      onMouseEnter={() => setShowButtons(true)}
      onMouseLeave={() => setShowButtons(false)}
    >
      <JsxParser
        bindings={setBindings(newItem)}
        components={{ ...Components.allComponentsAsJson(), Box, Stack }}
        jsx={template}
        renderInWrapper={false}
        onError={() => {}}
        renderError={({ error }) => <span>{error}</span>}
        showWarnings={false}
        disableKeyGeneration={true} // prevent provided jsx to be reconstructed at every render and to keep their states
        allowUnknownElements={false}
        renderUnrecognized={(tagName) => (
          <span style={{ color: theme.palette.primary.main }}>
            Error: unrecognized tag ({tagName})...
          </span>
        )}
      />
      {showButtons && 
        <Stack sx={buttonSx}>
          <IconButton
            id="deleteItemButton"
            aria-label="delete item" 
            color="inherit"
            onClick={() => handleDeleteItem(newItem[Globals.itemIdFieldName])}
          >
            <HighlightOffIcon />
          </IconButton>
        </Stack>
      }
    </Box>
  );
}

//export default memo(Item);
export default Item;
