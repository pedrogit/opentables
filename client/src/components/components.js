import React from "react";
import Stack from "@mui/material/Stack";
import Button from "@mui/material/Button";
import ButtonGroup from "@mui/material/ButtonGroup";
import Typography from "@mui/material/Typography";
import Link from "@mui/material/Link";
import Box from "@mui/material/Box";
import Popover from "@mui/material/Popover";
import TextField from "@mui/material/TextField";
import MUISelect from '@mui/material/Select';
import MenuItem from '@mui/material/MenuItem';
import ClickAwayListener from '@mui/material/ClickAwayListener';
import { useTheme } from "@mui/material/styles";
import ReCAPTCHA from "react-google-recaptcha";

import VisibilityPasswordTextField from "./VisibilityPasswordTextField";

const Globals = require("../common/globals");
const Errors = require("../common/errors");

/********************
 *  Label component
 ********************/
function Label({
  val, // label text
  vertical = false, // label is over the component (vertical=true), or  on the left size of it (vertical=false)
  nolabel = false, // do not dispay the label
  sx // label sx
}) {
  const theme = useTheme();
  var fontSize = vertical
    ? theme.typography.caption
    : theme.typography.body1;

  var defaultSx = {
    fontSize: fontSize,
    color: theme.palette.primary.main,
    fontWeight: "bold",
    marginRight: 1,
  };

  if (!nolabel) {
    var separator = vertical ? null : <>&nbsp;:</>;
    return (
      <Typography sx={{ ...defaultSx, ...sx }}>
        {val}
        {separator}
      </Typography>
    );
  }

  return null;
}

/********************
 *  Text component
 ********************/
function Text({
  val, 
  inline = false, // edit mode is inline 
  wrappedInform = false, // component is part of a form
  pretty = false, // make inline inputs pretty
  editmode = false, // switch between read and edit mode
  vertical = false, // vertical lavels (horizontal otherwise)
  label, // label
  nolabel = false, // do not display label
  labelSx = {}, // label sx
  sx  // component sx
}) {
  const propName = val ? (val.prop === undefined ? "Missing property name" : val.prop) : '';
  const propVal = val ? (val.val === undefined ? "Missing value" : val.val) : '';

  const valueRef = React.useRef();
  const theme = useTheme();

  const [editVal, setEditVal] = React.useState(propVal);
  const [isEditing, setIsEditing] = React.useState(editmode);

  // reset the editValue when the component is updated with a new val
  React.useEffect(() => {
    if (wrappedInform && val && val.val !== undefined) {
      setEditVal(val.val);
    }
  }, [val] );

  if (val && (propName || editVal)) {

    var defaultSx = {
      marginTop: inline && pretty && (editmode || isEditing) ? "8px" : "inherit",
      marginBottom: inline && pretty && (editmode || isEditing) ? "8px" : "inherit"
    };

    const handleEdit = (e) => {
      if (!wrappedInform) {
        val.handleItemAuth({
          action: 'patch', 
          propName: val.prop, 
          callback: (auth) => {
            if (!editVal) {
              setEditVal(propVal);
            }
            setIsEditing(true);
          }
        });
      }
    };

    const handleChange = (newVal) => {
      if (wrappedInform) {
        // change the List editingItem
        val.handleSaveProperty({ [propName]: newVal });
      }
      else {
        setEditVal(newVal);
      }
    };

    const handleSave = () => {
      if (!wrappedInform && editVal !== propVal) {
        val.handleSaveProperty({ [propName]: editVal }, (success, val) => {
          if (success) {
            setIsEditingOff();
            setEditVal(val[propName]);
          }
        });
      }
    };

    const keyPressed = (e) => {
      if (!wrappedInform) {
        if (e.keyCode === 13) { // enter
          handleSave();
        }
      }
      if (e.keyCode === 27) { // escape
        setEditVal(propVal);
        setIsEditingOff()
      }
    };

    const setIsEditingOff = () => {
      if (!wrappedInform) {
        setIsEditing(false);
        if (!editVal) {
          setEditVal(propVal);
        }
      }
    }

    const inputLabel = 'Edit "' + propName + '"...';
    const getWidth = () => {
      const labelFontWidth = 0.75 * theme.typography.fontSize * 0.4;
      const labelW = labelFontWidth * inputLabel.length;
      const inputPadding = 28;
      const popoverPadding = 16;
      return Math.max(
        (valueRef && valueRef.current && valueRef.current.offsetWidth ? valueRef.current.offsetWidth : 0) + inputPadding + popoverPadding,
        labelW + 2 * labelFontWidth + inputPadding + popoverPadding
      );
    };

    return (
      <>
        <Stack direction={vertical ? "column" : "row"}>
          {(!(inline && pretty) || !(editmode || isEditing)) && <Label 
            vertical={vertical} 
            val={label ? label : propName.charAt(0).toUpperCase() + propName.slice(1)}
            nolabel={nolabel} 
            sx={{...defaultSx, ...sx, ...labelSx}}
          />}
          {(editmode || isEditing) && inline ? (
            <>
            <ClickAwayListener onClickAway={setIsEditingOff}>
              <TextField
                name={propName}
                variant={pretty ? "outlined" : "filled"}
                sx={{ ...defaultSx, ...sx, backgroundColor: theme.palette.primary.palebg}}
                inputProps={pretty? {} : {
                  style: {padding: '0px'}
                }}
                size="small"
                label={pretty ? (label ? label : propName.charAt(0).toUpperCase() + propName.slice(1)) : null}
                fullWidth
                value={editVal}
                onChange={(e) => handleChange(e.target.value)}
                onKeyDown={(e) => keyPressed(e)}
                InputLabelProps={{shrink: true}}
                autoFocus={wrappedInform ? false : true}
              />
            </ClickAwayListener>
            </>
          ) : (
            <Typography
              id={propName}
              sx={{ ...defaultSx, ...sx }}
              onDoubleClick={handleEdit}
              ref={valueRef}
            >
              {propVal}
            </Typography>
          )}
        </Stack>
        <Popover
          open={(editmode || isEditing) && !inline}
          anchorEl={valueRef.current}
          onClose={setIsEditingOff}
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "left",
          }}
        >
          <Box sx={{ p: 1, width: getWidth() }}>
            <TextField
              sx={{backgroundColor: theme.palette.primary.palebg}}
              fullWidth
              id="outlined-basic"
              variant="outlined"
              size="small"
              label={inputLabel}
              value={editVal}
              onChange={(e) => handleChange(e.target.value)}
              onKeyDown={(e) => keyPressed(e)}
              autoFocus
            />
          </Box>
        </Popover>
      </>
    );
  }
  return null;
}

/********************
 *  Select component
 ********************/
 function Select({
  val,
  options,
  wrappedInform = false, // component is surrounded with a ItemWrapperForm form
  pretty = false, // make inline inputs pretty
  editmode = false, // init in edit mode
  vertical = false, // vertical lavels (horizontal otherwise)
  label, // label
  nolabel = false, // do not display label
  labelSx = {}, // label sx
  sx  // component sx
}) {
  const propName = val ? (val.prop === undefined ? "Missing property name" : val.prop) : undefined;
  const propVal = val ? (val.val === undefined ? "Missing value" : val.val) : undefined;

  const [editVal, setEditVal] = React.useState(propVal);
  const [isEditing, setIsEditing] = React.useState(editmode);

  if (val && (propName || editVal)) {

    var defaultSx = {
      marginTop: wrappedInform && pretty && (editmode || isEditing) ? "8px" : "inherit",
      marginBottom: wrappedInform && pretty && (editmode || isEditing) ? "8px" : "inherit"
    };

    const handleEdit = (e) => {
      if (!wrappedInform) {
        val.handleItemAuth({
          action: 'patch', 
          propName: val.prop, 
          callback: (auth) => {
            if (!editVal) {
              setEditVal(propVal);
            }
            setIsEditing(true);
          }
        });
      }
    };

    const handleChange = (newVal) => {
      if (!wrappedInform && newVal !== propVal) {
        val.handleSaveProperty({ [propName]: newVal }, (success, data) => {
          if (success) {
            setIsEditingOff();
            setEditVal(data[propName]);
          }
        });
      }
    };

    const keyPressed = (e) => {
      if (e.keyCode === 27) { // escape
        setEditVal(propVal);
        setIsEditingOff()
      }
    };

    const setIsEditingOff = () => {
      if (!wrappedInform) {
        setIsEditing(false);
        if (!editVal) {
          setEditVal(propVal);
        }
      }
    }

    return (
      <Stack direction={vertical ? "column" : "row"}>
        {(!(wrappedInform && pretty) || !(editmode || isEditing)) && <Label 
          vertical={vertical} 
          val={label ? label : propName.charAt(0).toUpperCase() + propName.slice(1)}
          nolabel={nolabel} 
          sx={{...defaultSx, ...sx, ...labelSx}}
        />}
        {(editmode || isEditing) ? (
          <>
            <MUISelect
              name={propName}
              open={editmode || isEditing}
              variant={"standard"}
              sx={{ ...defaultSx, ...sx}}
              MenuProps={{
                anchorOrigin: {
                  vertical: "bottom",
                  horizontal: "left"
                },
                transformOrigin: {
                  vertical: "top",
                  horizontal: "left"
                },
                getContentAnchorEl: null,
              }}
              SelectDisplayProps={{style: {padding:'0px 20px 0px 0px '}}}
              size="small"
              autoWidth={true}
              value={editVal}
              onChange={(e) => {handleChange(e.target.value)}}
              onKeyDown={(e) => keyPressed(e)}
              onClose={setIsEditingOff}
            >
              <MenuItem value="" dense={true}><em>unset</em></MenuItem>
              {options && options.map(opt => <MenuItem value={opt} dense={true}>{opt}</MenuItem>)}
            </MUISelect>
          </>
        ) : (
          <Typography
            id={propName}
            sx={{ ...defaultSx, ...sx }}
            onDoubleClick={handleEdit}
          >
            {propVal}
          </Typography>
        )}
      </Stack>
    );
  }
  return null;
}

/*************************
 *  Viewlink component
 *************************/
function Viewlink({
  text,  // text to display under the link
  viewid // viewid of the view to display
}) {
  return (
    <Link onClick={viewid && viewid.setViewId ? (() => (viewid.setViewId)(viewid.val)) : null}>
      {text ? (text.val ? text.val : "Text property missing...") : "Text property missing..."}
    </Link>
  )
}

/*************************
 *  Password component
 *  
 *  Always in edit mode
 *************************/
function Password({
  val,
  inline = true, // inline or popop (for now always inline)
  pretty = false, // make inline inputs pretty
  vertical = false, // label placement
  label, // label to display
  nolabel = false, // do not display any label
  labelSx = {}, // label sx
  sx // input sx
}) {
  const propName = val ? (val.prop === undefined ? "Missing property name" : val.prop) : undefined;
  const propVal = val ? (val.val === undefined ? "Missing value" : val.val) : undefined;

  const [editVal, setEditVal] = React.useState(propVal);
  const theme = useTheme();

  if (val && (propName || editVal)) {
    var defaultSx = {
      marginTop: inline && pretty ? "8px" : "inherit",
      marginBottom: inline && pretty ? "8px" : "inherit"
    };

    const handleChange = (newVal) => {
      setEditVal(newVal);
    };

    return (
      <>
        <Stack direction={vertical ? "column" : "row"}>
          {!pretty && <Label 
              vertical={vertical} 
              val={label ? label : propName.charAt(0).toUpperCase() + propName.slice(1)}
              nolabel={nolabel} 
              sx={{...defaultSx, ...sx, ...labelSx}}
            />
          }
          
          <VisibilityPasswordTextField
            name={propName}
            value={editVal}
            sx={{ ...defaultSx, ...sx, backgroundColor: theme.palette.primary.palebg}}
            inputProps={pretty? {} : {
              style: {padding: "0px"}
            }}
            variant={inline && !pretty ? "filled" : "outlined"}
            size="small"
            fullWidth
            label={inline && !pretty ? null : (label ? label : propName.charAt(0).toUpperCase() + propName.slice(1))}
            autoComplete="off"
            onChange={(e) => handleChange(e.target.value)}
            InputLabelProps={{shrink: true}}
          />
        </Stack>
      </>
    )
  }
  return null
}

/*************************
 *  Add Item Wrapper Form
 *  
 *  Add cancel (or add or reset) and Add (or Register) buttons 
 *  around the item template and set the template component
 *  properties to necessary values (wrappedInform, inline, pretty)
 *************************/
function ItemWrapperForm({
  handlers, 
  otherProps, 
  children
}) {
  const defChildProps = {
    wrappedInform: true, 
    inline: true, 
    pretty: true, 
    editmode: true
  }
  const [recaptchaResponse, setRecaptchaResponse] = React.useState('');
  const [submitButtonDisabled, setSubmitButtonDisabled] = React.useState(true);

  const recaptchaRef = React.useRef();
  var options = {
    cancelLabel: "Cancel",
    cancelAction: () => {
      resetForm();
      otherProps.setAddItem(false);
    },
    addLabel: "Add",
    addAction: () => otherProps.setAddItem(false)
  };

  if (otherProps.addItemMode === Globals.addWithPersistentFormNoItems) {
    options = {
      ...options,
      cancelAction: () => otherProps.backToMainView(Globals.viewOnAllViewViewId),
      addLabel: otherProps.addLabel ? otherProps.addLabel : "Save",
      addAction: () => otherProps.backToMainView(Globals.viewOnAllViewViewId),
      addMessage: {
        severity:"success",
        title: "Success!",
        text: "Your new item was added..."
      }
    }
    if (otherProps && otherProps.addMessageText) {
      options.addMessage.text = otherProps.addMessageText;
    }
    if (otherProps && otherProps.addMessageTitle) {
      options.addMessage.title = otherProps.addMessageTitle;
    }
  }

  const resetRecaptcha = () => {
    if (recaptchaRef && recaptchaRef.current) {
      setRecaptchaResponse('');
      recaptchaRef.current.reset();
    }
  }

  const resetForm = () => {
    resetRecaptcha();
    otherProps.resetEditingItem();
  }
  
  if (otherProps.addItemMode === Globals.addWithPersistentFormAndItems) {
    options = {
      cancelLabel: "Reset",
      cancelAction: () => resetForm(),
      addLabel: "Add",
      addAction: () => resetForm()
    }
  }

  const extractValues = (target) => {
    var item = {};
    for (var i = 0; i < target.length - 1; i++) {
      if (target[i].type !== "button" && 
          target[i].type !== "fieldset" && 
          target[i].name !== undefined && 
          target[i].defaultValue !== undefined &&
          target[i].defaultValue !== null &&
          target[i].defaultValue !== "") {
        item[target[i].name] = target[i].defaultValue;
      }
      if (target[i].name === Globals.gRecaptchaResponse && recaptchaResponse) {
        item[Globals.gRecaptchaResponse] = recaptchaResponse;
      }
    }
    return item;
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    if (!otherProps.recaptcha || (otherProps.recaptcha && recaptchaResponse)) {
      var newItem = extractValues(e.target);
      handlers.handleAddItem({
        item: newItem,
        addToLocalList: otherProps.addItemMode === Globals.addWithPersistentFormNoItems ? false : true,
        callback: (success) => {
          if (success) {
            setSubmitButtonDisabled(true);
            if (options.addMessage) {
              handlers.setErrorMsg(options.addMessage);
            }
            options.addAction();
          }
        }
      })
    }
    else {
      handlers.setErrorMsg({text: Errors.ErrMsg.Recaptcha_Failed});
    }
   }

  const childrenWithProp = React.Children.map(children, child => {
    if (React.isValidElement(child)) {
      return React.cloneElement(child, defChildProps);  
    }
    return child;
  });

  const keyPressed = (e) => {
    if (e.keyCode === 27) {
      options.cancelAction();
    }
    if (e.keyCode !== 13 && e.keyCode !== 8 && e.keyCode !== 46) {
      setSubmitButtonDisabled(false);
    }
    /*else {
      setSubmitButtonDisabled(true);
    }*/
  };

  return (
    <form
      onSubmit={handleSubmit}
      onKeyDown={keyPressed}
    >
      {childrenWithProp}
      <Stack
        direction={{ xs: 'column', sm: 'row' }}
        justifyContent={otherProps.recaptcha ? "space-between" : "flex-end"}
      >
        {otherProps.recaptcha &&
          <ReCAPTCHA
            id='g-recaptcha'
            ref={recaptchaRef}
            sitekey={window.Cypress 
                      ? '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI' // always positive key
                      : '6LcH-QkfAAAAAEKeUGIPbeY1OUlN4aQRkMyRoY_V'}
            onChange={(value) => setRecaptchaResponse(value)}
            onExpired={() => setRecaptchaResponse('')}
          />
        }
        <Stack 
          direction="row" 
          justifyContent="flex-end"
          alignItems="flex-end"
        >
          <ButtonGroup variant="contained" size="small">
            <Button id="addCancelItemFormButton" onClick={() => options.cancelAction()}>{options.cancelLabel}</Button>
            <Button id="addItemFormButton" type="submit" disabled={submitButtonDisabled}>{options.addLabel}</Button>
          </ButtonGroup>
        </Stack>
      </Stack>
    </form>
  )
}

function allComponentsAsJson() {
  return { Text, Select, Label, Viewlink, Password, ItemWrapperForm };
}

export { allComponentsAsJson, Text };
