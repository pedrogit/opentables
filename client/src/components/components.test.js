/* eslint-env mocha */
import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import { Text } from "./components";

describe("Text", () => {
  it("renders Text without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Text val={{ prop: "Prop name", val: "prop value" }} />,
      div
    );
  });

  it("renders Text without prop name without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Text val={{ val: "prop value" }} />, div);
  });

  it("renders Text without value without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Text val={{ prop: "Prop name" }} />, div);
  });

  it("renders Text without prop without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Text />, div);
  });

  it("has a valid snapshot", () => {
    const component = renderer.create(
      <Text val={{ prop: "Prop name", val: "prop value" }} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
