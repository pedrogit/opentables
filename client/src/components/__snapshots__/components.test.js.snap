// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text has a valid snapshot 1`] = `
<div
  className="css-m69qwo-MuiStack-root"
>
  <p
    className="MuiTypography-root MuiTypography-body1 css-o6haml-MuiTypography-root"
  >
    Prop name
    Â :
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-199lc0q-MuiTypography-root"
    id="Prop name"
    onDoubleClick={[Function]}
  >
    prop value
  </p>
</div>
`;
